### PR TITLE
Add test suites for `core:sync` and `core:sync/chan`

### PR DESCRIPTION
**Pull request by Feoramund**
This patch required much tinkering to get right. In the process, I've fixed over a dozen synchronization bugs, including the failure to detect more than 1 core (!) on FreeBSD and NetBSD in the core library and (for FreeBSD only) the Odin compiler itself. That bugfix itself improves the test runner's speed on those two platforms.

Of note, I converted @laytan's hack in the POSIX thread startup proc to use a semaphore instead of a mutex/condition variable, which should be more robust since it depends on a non-zero state instead of needing to loop on the condition.

I took to the entire test suite with `-sanitize:thread` and eliminated every data race I could find, and with the test suite now testing every primitive, I was able to eliminate several deadlocks. I rewrote the recursive benaphore because I was having trouble avoiding deadlocks with the particular order it took to modifying state. I think it's more robust now, but I'm not 100% certain.

I've run these tests locally on Linux, FreeBSD, and NetBSD several hundred times, with and without `-sanitize:thread` on (for Linux only, of course). Given the nature of sync issues, it's necessary to run it multiple times just to be sure.

Unfortunately, one of the `core:sync` tests isn't completing on the Darwin CI, which stalls the entire test runner because (I think) that the runner is having an issue with triggering the failure timeout via pthread cancel, which we've discussed previously about its unreliability on Darwin. I'm not sure what the actual underlying issue would be though, with regard to which test or tests are causing a deadlock. I do not have a local Darwin machine to investigate, but if someone could take a look at it, that'd be great. For now, I've marked the test suite as disabled on Darwin.

To summarize, we now have basic coverage for every `core:sync` primitive as well as `chan.Chan`. There may still be some edge cases for `chan`, but I fixed the ones I could find.

### CODE REVIEW

**Comment by Feoramund** [comment]
Regarding the NetBSD CI failure:
```
[ERROR] --- [2024-09-11 19:10:43] [structs.odin:66:wait_for()] waitpid() failure: Interrupted system call
```
I saw this come up intermittently before in my repo's CI, and it only started happening after I fixed the CPU count detection in `core`. I haven't been able to replicate it locally, and I am not sure yet what the fix would be.

**Comment by laytan** [comment]
>  I haven't been able to replicate it locally, and I am not sure yet what the fix would be.

Ah looks like I wrote that wrong, the loop should check for EINTR and loop (continue).

**Comment by Feoramund** [comment]
Regarding the stalls on Darwin, I want to mention that I tried porting the usage of the `SHARED` flags that were introduced into the compiler by 03426175ae, into the usage of `ulock_*` and `os_sync_*` in the Odin core library as a possible fix to my repo's CI, but it continued to stall.

**Comment by flysand7** [comment]
> I've run these tests locally on Linux, FreeBSD, and NetBSD several hundred times, with and without -sanitize:thread on (for Linux only, of course). Given the nature of sync issues, it's necessary to run it multiple times just to be sure.

Just a heads up, not sure about FreeBSD and NetBSD kernels, but on linux sometimes running multiple times does not help, since the scheduler seemed very deterministic. I remember when debugging some of my multithreaded code, if it occurred once it was ocurring every single time, and if it never occurred I had trouble triggering it.

Not sure what the solution is, except running a shitton of times, while inserting calls to `sched_yield` at some random places (usually before taking or releasing locks). But yeah It's just a thing I've noticed, maybe since 5.0 linux did change their scheduler or something, I'm not sure.

File: core/sync/extended.odin [hunk 1]
```diff
@@ -494,14 +495,14 @@ for other threads for entering.
...
 	recursion := b.recursion
+
 	if recursion == 0 {
-		b.owner = 0
-	}
-	if atomic_sub_explicit(&b.counter, 1, .Release) > 0 {
-		if recursion == 0 {
+		if atomic_sub_explicit(&b.counter, 1, .Relaxed) == 1 {
+			atomic_store_explicit(&b.owner, 0, .Release)
+		} else {
```
Comment (flysand7) [inline core/sync/extended.odin:505]: Pretty sure you forgot to reset the owner to `0`, in the other branch of this if statement.

File: core/sync/chan/chan.odin [hunk 2]
```diff
@@ -164,12 +164,17 @@ send_raw :: proc "contextless" (c: ^Raw_Chan, msg_in: rawptr) -> (ok: bool) {
 	}
 	if c.queue != nil { // buffered
 		sync.guard(&c.mutex)
-		for c.queue.len == c.queue.cap {
+		for !sync.atomic_load(&c.closed) &&
```
[the remark this answers is not in the document]
Comment (Feoramund) [reply core/sync/chan/chan.odin:167]: On second look, we might be able to change all `c.closed` loads to non-atomic, not just this one. I was following convention on this change, since the other accesses are atomic.

File: core/sync/extended.odin (hunk 1 above)
Comment (Feoramund) [reply core/sync/extended.odin:505]: By virtue of the semaphore post signalling the other waiting thread to unwait at line 457, the next instruction would set the appropriate owner variable. It's only necessary to set the owner if no one else is going to.

File: core/sync/chan/chan.odin (hunk 2 above)
[the remark this answers is not in the document]
Comment (flysand7) [reply core/sync/chan/chan.odin:167]: Yeah, this should save an extra `lwsync` operation on ARM, iirc

[the remark this answers is not in the document]
Comment (Feoramund) [reply core/sync/chan/chan.odin]: In the end, I removed all atomic loads and stores, as well as the other two mutexes. Everything is under guard by `mutex` already.

**Comment by laytan** [comment]
I can look into the Darwin issues with this.

**Review by gingerBill — APPROVED** [review]
